### PR TITLE
feat(lua): add vim.resolve_bufnr

### PR DESCRIPTION
**Pull request by glepnir**
Problem: there has lots of duplicate codes about detect current bufnr with nil. and also lot of similar code at the plugin side.

Solution: add a generic vim.resolve_bufnr in shared.

### CODE REVIEW

**Comment by clason** [comment]
I disagree; most of these cases should be handled by consistent convention (`bufnr` is mandatory with `0` for current). We should _never_ use `nil` for current in our codebase.

**Comment by mfussenegger** [comment]
This would violate the documented `dev-pattern`:

> - When accepting a buffer id, etc., 0 means "current buffer", nil means "all
> buffers".  Likewise for window id, tabpage id, etc.
> - Examples: |vim.lsp.codelens.clear()| |vim.diagnostic.enable()|

**Comment by lewis6991** [comment]
A nice cleanup and I agree this doesn't need to be exposed.

File: runtime/lua/vim/lsp/codelens.lua [hunk 1]
```diff
@@ -104,16 +104,12 @@ function M.run()
...
-  return bufnr == 0 and api.nvim_get_current_buf() or bufnr
-end
-
 --- Clear the lenses
 ---
 ---@param client_id integer|nil filter by client_id. All clients if nil
 ---@param bufnr integer|nil filter by buffer. All buffers if nil, 0 for current buffer
 function M.clear(client_id, bufnr)
-  bufnr = bufnr and resolve_bufnr(bufnr)
+  bufnr = bufnr and vim.resolve_bufnr(bufnr)
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/codelens.lua:112]: This would change semantics and break it

**Review by mfussenegger — CHANGES_REQUESTED** [review]
Just looked at the changes in more detail and it appears that this would break some behaviour. 

In other callsites the bufnr is already non-nil.

File: runtime/lua/vim/lsp/codelens.lua (hunk 1 above)
Comment (glepnir) [reply runtime/lua/vim/lsp/codelens.lua:112]: This is easy to fix I didn't take a quick look at the code here

**Comment by justinmk** [comment]
I mentioned in other discussions that we can start by collecting these common patterns in an internal-only `vim._util` module. Obviously this should not be a top-level `vim.resolve_bufnr` function.

**Pull request closed without merging**